awswrangler/s3/_read.py

```diff
@@ -116,3 +136,183 @@ def _check_version_id(
     return (
         version_id if isinstance(version_id, dict) else {paths[0]: version_id} if isinstance(version_id, str) else None
     )
+
+
+class _InternalReadTableMetadataReturnValue(NamedTuple):
+    columns_types: Dict[str, str]
+    partitions_types: Optional[Dict[str, str]]
+    partitions_values: Optional[Dict[str, List[str]]]
+
+
+class _TableMetadataReader(ABC):
+    @abstractmethod
+    def read_schemas_from_files(
```

Comment:
I think we can also have common implementation of the schema reads here as it seems the only thing that is different between parquet and orc is the downstream function that is submitted to the executor (that reads a single schema from the path). 

Can we do same thing here as with ray datasources for example - just override `_read_block` or in this case `_read_metadata_file` and keep the rest the same?